Swap GA graduation: blog-post (#48099)
kubernetes/website: content/en/blog/_posts/2024-09-25-swap-ga-graduation/2024-09-25-swap-ga-graduation.md

Reviewer: If there is no SVG to reference, filename should be similar:
`content/en/blog/_posts/2024-09-25-linux-swap-ga.md` (adjust once date is assigned by Release Comms)

Author: Right! Thanks!

On Linux nodes, Kubernetes only supports running with swap enabled for hosts that use cgroup v2.
On cgroup v1 systems, all Kubernetes workloads are not allowed to use swap memory.

## Install a swap-enabled cluster with kubeadm

### Before you begin

It is required for this demo that the kubeadm tool be installed, following the steps outlined in the
[kubeadm installation guide](/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm).
If swap is already enabled on the node, cluster creation may proceed.
If swap is not enabled, please refer to the provided instructions for enabling swap.

### Create a swap file and turn swap on

I'll demonstrate creating 4GiB of swap, both in the encrypted and unencrypted case.

#### Setting up unencrypted swap

An unencrypted swap file can be set up as follows.

```bash
dd if=/dev/zero of=/swapfile bs=128M count=32
chmod 600 /swapfile
mkswap /swapfile
swapon /swapfile
```

#### Setting up encrypted swap

An encrypted swap file can be set up as follows.
Bear in mind that this example uses the `cryptsetup` binary (which is available
on most Linux distributions).

```bash

> dd if=/dev/zero of=/swapfile bs=128M count=32
> chmod 600 /swapfile
> cryptsetup --type plain --cipher aes-xts-plain64 --key-size 256 -d /dev/urandom open /swapfile cryptswap
> mkswap /dev/mapper/cryptswap
> swapon /dev/mapper/cryptswap

Reviewer: How about:
```suggestion
# Allocate storage and restrict access
fallocate --length 4GiB /swapfile
chmod 600 /swapfile

# Create an encrypted device backed by the allocated storage
cryptsetup --type plain --cipher aes-xts-plain64 --key-size 256 -d /dev/urandom open /swapfile cryptswap

# Format the swap space
mkswap /dev/mapper/cryptswap

# Activate the swap space for paging
swapon /dev/mapper/cryptswap
```

```

#### Verify that swap is enabled

Swap can be verified to be enabled with both `swapon -s` command or the `free` command

```shell
> swapon -s
Filename				Type		Size		Used		Priority
/dev/dm-0                               partition	4194300		0		-2
```

```shell
> free -h
               total        used        free      shared  buff/cache   available
Mem:           3.8Gi       1.3Gi       249Mi        25Mi       2.5Gi       2.5Gi

> Swap:          4.0Gi          0B       4.0Gi

Reviewer: We should separate the commands from output; everything marked as shell should be valid POSIX shell (or close).

```

#### Enable swap on boot

After setting up swap, to start the swap file at boot time,

> add a line similar to `/swapfile swap swap defaults 0 0` to `/etc/fstab` file.

Reviewer: ```suggestion
you either set up a systemd unit to activate (encrypted) swap, or you
add a line similar to `/swapfile swap swap defaults 0 0` into `/etc/fstab`.
```
?

Using fstab isn't deprecated, but it is a legacy approach.

the container runtime is then responsible for writing these settings to the container-level cgroup.

## How can I monitor swap?

### Node-level metric statistics

Kubelet now collects
[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
which can be accessed at the `/metrics/resource` (which is used mainly by monitoring
tools like Prometheus) and `/stats/summary` (which is used mainly by Autoscalers) kubelet HTTP endpoints.
This allows clients who can directly interrogate the kubelet to
monitor swap usage and remaining swap memory when using `LimitedSwap`.
Additionally, a `machine_swap_bytes` metric has been added to cadvisor to show
the total physical swap capacity of the machine.

### NFD (Node Feature Discovery)

[Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery)
is a Kubernetes addon for detecting hardware features and configuration.
It can be utilized to discover which nodes are provisioned with swap.

As an example, to figure out if a certain node is provisioned with swap,
the following command can be executed while replacing `<NODE-NAME>` with the node's name:
```shell

> > k get node <NODE-NAME> -o yaml | grep "feature.node.kubernetes.io/memory-swap"

Reviewer: If we can use `jsonpath` output format rather than rely on `grep`, even better. This is informal but I'm minded to teach something that's 100% reliable. People might infer that this code is what they should put in scripts.

Also, what if the node is ever annotated `feature.node.kubernetes.io/memory-swap: "false"`?

Linux versions as well.

In order to verify whether the node supports the `noswap` option, the kubelet will do the following:
* If the kernel's version is above 6.3 then the `noswap` option will be assumed to be supported.
* Otherwise, kubelet would try to mount a dummy tmpfs with the `noswap` option at startup.
  If kubelet fails with an error indicating of an unknown option, `noswap` will be assumed
  to not be supported, hence will not be used.
  A kubelet log entry will be emitted to warn the user about memory-backed volumes might swap to disk.
  If kubelet succeeds, the dummy tmpfs will be deleted and the `noswap` option will be used.
  * If the `noswap` option is not supported, kubelet will emit a warning log entry,
    then continue its execution.

It is deeply encouraged to encrypt the swap space.
See the [section above](#setting-up-encrypted-swap) with an example for setting unencrypted swap.
However, handling encrypted swap is not within the scope of kubelet;
rather, it is a general OS configuration concern and should be addressed at that level.
It is the administrator's responsibility to provision encrypted swap to mitigate this risk.

### Evictions

> Since the NodeSwap feature introduces basic swap support and leaves more advanced topics to
> follow-up features, swap-based evictions are out-of-scope for this feature.

Reviewer: When this blog article is published, the NodeSwap feature gate will have no effect. So:
```suggestion
At the moment Kubernetes only has basic swap support for Linux nodes. For Kubernetes v1.32,
the kubelet doesn't ever use pressure on swap space as a signal for eviction.
```

With swap being disabled, the memory eviction threshold should be set in a way that will let kubelet's
eviction manager identify memory spikes in time so it can execute custom logic and take care of reclaiming
memory and evicting pods before the kernel's OOM killer is invoked.

However, with swap enabled, the situation is fundamentally different.
Since swapping is a heavy operation in terms of performance, the kernel will try avoiding
it as long as it can, and will usually start swapping only when it lacks physical memory.

Therefore, when swap is enabled, it is acceptable to use all the available RAM which
will result in the node using some of the swap memory.

#### Current eviction limitations

> Because evictions are out-of-scope for this feature, the eviction manager would remain swap-unaware.

Reviewer: Is this true:
```markdown
For both Linux and Windows nodes, the eviction manager is not aware of how much of the swap storage is used.
```
?

Author: Yes it is. I'll update this.
See my comment regarding Windows [here](https://github.com/kubernetes/website/pull/48099#issuecomment-2381288665).

Let X be the memory threshold below which the kernel initiates swapping.
This means that most probably either one of the two behaviors will take place:
1. Memory evictions threshold > X: In this scenario, kubelet's eviction manager would kick in before the kernel has
   a chance to swap memory.
   This means that the kernel would generally not be able to swap memory (unless the memory spike
   is fast enough for the eviction manager to handle).
2. Memory evictions threshold <= X: In this scenario, the kernel would start swapping memory before the memory threshold 
   would be met. This practically means that the eviction manager would never kick in, hence it is practically turned off (unless
   for extreme cases where the memory spike is faster than the kernel's ability to swap memory). 

The cluster admin can choose which of the above is the desired behavior.

> An advanced cluster admins can use the approach outlined [below](#advanced-best-practices-for-manually-setting-memory-evictions).

Reviewer: ```suggestion
An advanced cluster admin can use the approach outlined [below](#advanced-best-practices-for-manually-setting-memory-evictions).
```

#### Advanced good practice for manually setting memory evictions

> Simplistically, the kernel uses three "watermarks" (i.e. thresholds),

Reviewer: ```suggestion
Simplistically, the kernel uses three thresholds (known as _watermarks_),
```

which are called "high", "min", and "low", such that high > min > low.
When the "min" threshold is met, the kernel enters an "indirect reclaim" state, which basically means that the `kswapd`
daemon becomes active and asynchronously starts to reclaim memory.
If the "low" threshold is being met, the kernel enters a "direct reclaim" state, in which `kswapd` would aggressively
reclaim memory and memory would be throttled for most of the processes.
Later, when the "min" threshold is reached, the kernel goes back to "indirect reclaim", then when the "high" threshold
is met the node is not considered under pressure anymore.

> ![swap_watermarks](swap-watermarks.png)

Author: Changed to "Linux kernel swap memory flow". WDYT?

is a Kubernetes addon for detecting hardware features and configuration.
It can be utilized to discover which nodes are provisioned with swap.

As an example, to figure out which nodes are provisioned with swap,
use the following command:
```shell

> > kubectl get nodes -o jsonpath='{range .items[*]}{.metadata.name}: {.metadata.labels.feature\.node\.kubernetes\.io/memory-swap}{"\n"}{end}'

Reviewer: ```suggestion
kubectl get nodes -o jsonpath='{range .items[*]}{.metadata.name}: {.metadata.labels.feature\.node\.kubernetes\.io/memory-swap}{"\n"}{end}'
```

Reviewer: Extra credit: handle missing annotations.

Author: Done + the extra credit :wink:
